drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,470 @@
+	 * but for simplicity we will only use 1 per speed mode.
+	 */
+
+	/* Speed timer
+	 * 0: High speed timer
+	 * 1: Low speed timer
+	 */
+	struct pwm_led_esp32_timer timer_cfg[2];
+
+	struct pwm_led_esp32_gpio_speed_mode_range gpio_range_hs;
+	struct pwm_led_esp32_gpio_speed_mode_range gpio_range_ls;
+
+	const struct pwm_led_esp32_peripheral peripheral;
+};
+
+/* TODO: Remove these functions after this PR:
+ * https://github.com/zephyrproject-rtos/zephyr/pull/5113
+ */
+static inline void set_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) | v, mem_addr);
+}
+
+static inline void clear_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
+}
+
+static void esp32_enable_peripheral(
+	const struct pwm_led_esp32_peripheral *peripheral)
+{
+	set_mask32(peripheral->clk, DPORT_PERIP_CLK_EN_REG);
+	clear_mask32(peripheral->rst, DPORT_PERIP_RST_EN_REG);
+}
+
+static const char *esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+/* end Remove after PR 5113 */
+
+static void pwm_led_esp32_low_speed_update(int speed_mode, int channel)
+{
+	u32_t reg_addr;
+
+	if (speed_mode == PWM_LED_ESP32_LOW_SPEED) {
+		reg_addr = LEDC_LSCH0_CONF0_REG + 14*channel;
+		sys_set_bit(reg_addr, LEDC_PARA_UP_LSCH0_S);
+	}
+}
+
+static void pwm_led_esp32_update_duty(int speed_mode, int channel)
+{
+	volatile u32_t conf0_addr;
+	volatile u32_t conf1_addr;
+	int offset_addr = 14*channel;
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		conf0_addr = LEDC_HSCH0_CONF0_REG + offset_addr;
+		conf1_addr = LEDC_HSCH0_CONF1_REG + offset_addr;
+	} else {
+		conf0_addr = LEDC_LSCH0_CONF0_REG + offset_addr;
+		conf1_addr = LEDC_LSCH0_CONF1_REG + offset_addr;
+	}
+
+	sys_set_bit(conf0_addr, LEDC_SIG_OUT_EN_LSCH0_S);
+	sys_set_bit(conf1_addr, LEDC_DUTY_START_LSCH0_S);
+
+	pwm_led_esp32_low_speed_update(speed_mode, channel);
+}
+
+static void pwm_led_esp32_duty_config(int speed_mode,
+				      int channel,
+				      int duty_val,
+				      bool duty_direction,
+				      u32_t duty_num,
+				      u32_t duty_cycle,
+				      u32_t duty_scale)
+{
+	volatile u32_t hpoint_addr;
+	volatile u32_t duty_addr;
+	volatile u32_t conf1_addr;
+	volatile u32_t conf1_val;
+
+	int offset_addr = 14*channel;
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		hpoint_addr = LEDC_HSCH0_HPOINT_REG + offset_addr;
+		duty_addr = LEDC_HSCH0_DUTY_REG + offset_addr;
+		conf1_addr = LEDC_HSCH0_CONF1_REG + offset_addr;
+	} else {
+		hpoint_addr = LEDC_LSCH0_HPOINT_REG + offset_addr;
+		duty_addr = LEDC_LSCH0_DUTY_REG + offset_addr;
+		conf1_addr = LEDC_LSCH0_CONF1_REG + offset_addr;
+	}
+
+	sys_write32(0, hpoint_addr);
+
+	conf1_val = (duty_direction &   0x1) << 31 |
```

Comment:
Please use the `..._S` macros provided in ESP-IDF.  (After quickly looking around, I saw constants like `LEDC_DUTY_START_HSCH0_S`, `LEDC_DUTY_NUM_HSCH0_S`, `LEDC_DUTY_CYCLE_HSCH0_S`.)

Nitpick: avoid lining up things in statements like this.

Reply:
I used hardcoded because of the 80 lines, but I'll try another way.

Reply:
Instead of using `bool duty_direction, u32_t duty_num, u32_t duty_cycle, u32_t duty_scale`
I'll use just one struct (u32_t) with these variables, saving up memory and avoiding the over 80 chars issue.